unstoppabledomains/UI-Kit
# [ECUX-313] Initiate NPM package

## Background

This PR introduces a public NPM package containing shared UI components used across the UD website codebase. As a first step, it re-exports our theme and all MUI components, hooks, types, icons, and colors to be name imported in the UD client like this:

```typescript
import {Typography, useMediaQuery} from '{@mui/material => @unstoppabledomains/ui-kit}';
import {makeStyles, withStyles, useStyles} from '{styles => @unstoppabledomains/ui-kit/styles}';
import {lightTheme, MAX_PAGE_CONTENT_WIDTH} from '{styles/theme => @unstoppabledomains/ui-kit/styles/theme}';
import {ExpandMore} from '{@mui/icons-material => @unstoppabledomains/ui-kit/icons}';
import {LoadingButton} from '{@mui/lab => @unstoppabledomains/ui-kit/lab}';
import {indigo as indigoColor} from '{@mui/material/colors => @unstoppabledomains/ui-kit/colors}';
```

Basically, this is the new imports mapping that would exist in the UD if this version is merged and published:

```json
{
  "@mui/material": "unstoppabledomains/ui-kit",
  "@mui/lab": "@unstoppabledomains/ui-kit/lab",
  "@mui/icons-material": "@unstoppabledomains/ui-kit/icons",
  "@mui/material/colors": "@unstoppabledomains/ui-kit/colors",
  "styles": "@unstoppabledomains/ui-kit/styles",
  "styles/theme": "@unstoppabledomains/ui-kit/styles/theme"
}
```

The reason why there're separate paths for @mui/lab, @mui/material, and @mui/icons-material re-exports, and not just an option to import everything directly from `@unstoppabledomains/ui-kit` root is that they all have duplicate exports:

<img width="1512" alt="Screenshot 2022-09-09 at 14 57 43" src="https://user-images.githubusercontent.com/10829201/189345073-72c8d6c3-ec4a-46f3-bfa4-e8a45fdacdbf.png">

<img width="1512" alt="Screenshot 2022-09-09 at 14 57 57" src="https://user-images.githubusercontent.com/10829201/189345087-2b3772c3-0caa-4c3b-bfb5-0a000cc16c8b.png">

A separate path for importing `@mui/material/colors` exists for distinguishing the colors from other re-exports, though it could be moved to `components` path, for example.

The initial version is set to `0.1.0`. According to [semver docs](https://semver.org/spec/v0.1.0.html#spec-item-5):

- Major version zero (0.y.z) is for initial development. Anything may change at any time. The public API should not be considered stable.
- Version 1.0.0 defines the public API. The way in which the version number is incremented is now dependent on this public API and how it changes.

## Related Issue

[ECUX-313](https://linear.app/unstoppable-domains/issue/ECUX-313/create-ud-ui-kit-public-npm-package)

## How to install

- Clone this repo: `git clone https://github.com/unstoppabledomains/UI-Kit`
- Checkout the working branch: `git checkout dominic/ecux-313-create-ud-ui-kit-public-npm-package`
- Run `yarn` at the project's root

## How to build

Run `yarn build` at the root to create a `dist` folder which will go to the UD `node_modules` after installing the package. It also builds types (put under `dist` alongside the `.js` files) and storybook static files.

## How to build and view storybook

- Run `yarn storybook` to see an introductory Storybook page at localhost:6006
- Run `yarn build-storybook` to generate static Storybook HTML inside `storybook-static` folder
- For now it doesn't contain any stories, but UD common components will get their own stories over time

## How to check bundle size difference

1. Inside the UD website repo, run a production build locally:
  - cd `client`
  - update `compress` flag to `true` inside `next.config.js`
  - build: `APP_ENV=production NEXT_PUBLIC_API_BASE_URL=http://localhost:8080/ NEXT_PUBLIC_CLIENT_URL=http://localhost:3000/ NEXT_PUBLIC_GAE_VERSION=UNKONWN NEXT_PUBLIC_APP_ENV=production NEXT_PUBLIC_DEPLOYMENT=production yarn build`
  - run build: `APP_ENV=production NEXT_PUBLIC_API_BASE_URL=http://localhost:8080/ NEXT_PUBLIC_CLIENT_URL=http://localhost:3000/ NEXT_PUBLIC_GAE_VERSION=UNKONWN NEXT_PUBLIC_APP_ENV=production NEXT_PUBLIC_DEPLOYMENT=production yarn start -p 3000`
2. Head over to a UD website page, open up the Network tab and observe the `_next/static` JS file sizes
3. Checkout `dominic/re-target-mui-imports` branch inside the website repo (see changes in [this PR](https://github.com/unstoppabledomains/unstoppable-domains-website/pull/5672))
4. Repeat steps 1 and 2

_Screenshots of next static file sizes are provided in the website [PR](https://github.com/unstoppabledomains/unstoppable-domains-website/pull/5672)_

## To Do

- [x] Fill out the `README` file with the following instructions:
  - [x] installation
  - [x] contributing
  - [x] running and building storybook
  - [x] building the package
- [x] Make sure it's possible to replace all `@mui/material` appearances to `@unstoppabledomains/ui-kit` across UD client codebase without breaking the type checks (also do smoke tests)
- [ ] <del>Set up the build process so that the default MUI imports inside the website also work (could be optional since it's a good idea to refactor all default MUI imports to named ones)</del> (not relevant anymore as the default imports have been refactored)
- [ ] <del>Re-export one styled UD component as a sample</del> (should probably be postponed to a subsequent PR)
- [x] Get rid of `dist` in the import paths

Notes for future improvements:
- Configure the UI Kit to re-export all types that exist under sub-paths, e.g. `import type {OverridableComponent} from '@mui/material/OverridableComponent';`
- For these two cases ([one](https://github.com/unstoppabledomains/unstoppable-domains-website/blob/master/client/components/developers/DocCards/cardList.tsx#L2) and [two](https://github.com/unstoppabledomains/unstoppable-domains-website/blob/master/client/components/common/Pagination.tsx#L9)), put MUI types usage back

## Additional deployment steps

Publish the package to be publicly available as `@unstoppabledomains/ui-kit@0.1.0`
